models: tolerate empty field names when unpacking binary points

jonseymour opened the pull request:
Prior to #5697 influx crashed with a panic if a binary point contained an empty field name. After #5697, influx entered an infinite loop in the same circumstance. With this commit, we silently elide any field with an empty field name when unmarshaling a binary point. A related PR (#5714) prevents SELECT INTO trying to create binary points with empty field names.

jonseymour commented:
/cc @jwilder @e-dard

jwilder commented:
This isn't quite the right way to fix this.  A point should never contain fields without a name. 
* `NewPoint` should ensure that fields passed in do not have any empty field names and return an error if they do.
* `AddField` should be removed since points should be created fully via `NewPoint`.
* select queries with math that produces columns without a name should not happen.  It needs to generate a column name if an alias is not used.

jonseymour commented:
Ok, np. I'll fix accordingly.

jonseymour commented:
@jwilder - should I trim the field name of leading and trailing whitespace before testing for equality with the empty string?

jonseymour commented:
@jwilder - I have implemented the first two suggestions with this series. I'll investigate what's required to fix the queries and may submit that as a separate series. I am going to leave 930fc42 in the series because the existing code didn't handle the case gracefully (it should have scanned to the end of the value buffer before skipping to the next field).

I'll check the result of the CI build later in the night.

e-dard commented:
@jonseymour thanks for this. I'm down with the following changes: (1) check for empty fields in `NewPoint`; (2) removing  `AddFields`.

Can you explain why 930fc42 needs to be accepted though?  `newFieldsFromBinary` is only called when retrieving the fields of a `Point` created via `NewPoint`, and `NewPoint` will now check for empty fields.

jonseymour commented:
@e-dard - the algorithm is just wrong - if the data is corrupt, the existing algorithm an infinite loop is guaranteed, as illustrated by the test cases that begin the series.. With the version I propose, it is guaranteed that the loop will terminate.

e-dard commented:
@jonseymour which test case are you referring to? Could you throw a link my way? Cheers!

jonseymour commented:
This commit 23be938. A transcript of executing this test, merged into both #5697 (88937ab0f)  and the commit immediately preceding (b8bb32e) is as follows:

```
greywedge2:influxdb jonseymour (1)$ cd models
greywedge2:models jonseymour (1)$ git checkout 88937ab0f
HEAD is now at 88937ab... Fixes #5664
greywedge2:models jonseymour (1)$ git merge --no-commit 23be938
Automatic merge went well; stopped before committing as requested
greywedge2:models jonseymour (1)$ go test ./...
--- FAIL: TestAddFieldWithEmptyName (1.00s)
	points_test.go:1790: failed: probable infite loop
FAIL
FAIL	github.com/influxdb/influxdb/models	1.021s
greywedge2:models jonseymour (1)$ git reset --hard HEAD
HEAD is now at 88937ab Fixes #5664
greywedge2:models jonseymour (1)$ git checkout 88937ab0f^1
Previous HEAD position was 88937ab... Fixes #5664
HEAD is now at b8bb32e... Merge pull request #5470 from influxdata/ross-packaging-updates
greywedge2:models jonseymour (1)$ git merge --no-commit 23be938
Updating b8bb32e..23be938
Fast-forward
 models/points_test.go | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)
greywedge2:models jonseymour (1)$ go test ./...
panic: runtime error: index out of range

goroutine 77 [running]:
github.com/influxdb/influxdb/models.scanTo(0x82049e930, 0x8, 0x10, 0x0, 0x82049663d, 0x0, 0x0, 0x82049e920, 0x1)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points.go:892 +0xa7
github.com/influxdb/influxdb/models.newFieldsFromBinary(0x82049e930, 0x8, 0x10, 0x0)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points.go:1396 +0x108
github.com/influxdb/influxdb/models.(*point).unmarshalBinary(0x82049cb40, 0xffffffffffffffff)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points.go:1298 +0x40
github.com/influxdb/influxdb/models.(*point).Fields(0x82049cb40, 0x82049e930)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points.go:1211 +0x3a
github.com/influxdb/influxdb/models.(*point).AddField(0x82049cb40, 0x1c7d60, 0x1, 0x15a540, 0x82049e940)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points.go:1217 +0x25
github.com/influxdb/influxdb/models_test.TestAddFieldWithEmptyName.func1(0x82049cab0, 0x82049a3c0)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points_test.go:1784 +0x35b
created by github.com/influxdb/influxdb/models_test.TestAddFieldWithEmptyName
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points_test.go:1786 +0x6c

goroutine 1 [chan receive]:
testing.RunTests(0x216df8, 0x2b3980, 0x48, 0x48, 0x101)
	/usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0x820355ef8, 0x820392280)
	/usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:494 +0x70
main.main()
	github.com/influxdb/influxdb/models/_test/_testmain.go:218 +0x116

goroutine 76 [select]:
github.com/influxdb/influxdb/models_test.TestAddFieldWithEmptyName(0x82049cab0)
	/Users/jonseymour/ninja/go/src/github.com/influxdb/influxdb/models/points_test.go:1787 +0x16a
testing.tRunner(0x82049cab0, 0x2b4028)
	/usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:456 +0x98
created by testing.RunTests
	/usr/local/Cellar/go/1.5.3/libexec/src/testing/testing.go:561 +0x86d
FAIL	github.com/influxdb/influxdb/models	0.017s
 
```

e-dard commented:
OK cool, I see. But that test is now irrelevant because you're removing `AddFields` and also doing the check for empty field names inside of `NewPoint`.

With the changes you have made to `NewPoint` it should no longer be possible to get to `newFieldsFromBinary` with an empty field name.

I think 930fc42 can be dropped from the PR. Hope that makes sense?

jonseymour commented:
As I said, i can create a test case with points.NewPointFromBytes (which does not involve any calls to the now deleted points.AddField) method which will force influxd into an infinite loop if you really want me to prove that the algorithm prior to 930fc42 is vulnerable.

jonseymour commented:
@e-dard I've created a test case which shows that points.NewPointFromBytes is vulnerable to an infinite loop if corrupt data is read from the binary encoding of the point. 

Given that it costs to little to protect the system from an infinite loop caused by a data corruption, there seems to be no sound reason not to accept (8490c6e) (which replaces 930fc42).

jonseymour commented:
To see the vulnerablity, checkout 9b56025 which does not include the fix and then run the tests in the model package.

e-dard commented on models/points_test.go:
```diff
@@ -1772,3 +1772,24 @@ t159,label=another a=2i,value=1i 1`
+
+func TestNewPointsRejectsEmptyFieldNames(t *testing.T) {
+	if _, err := models.NewPoint("foo", nil, models.Fields{"": 1}, time.Now()); err == nil {
+		t.Fatalf("expected error if field with empty field name is created - empty string")
```
nit: the failure message could be something like `got nil, expected error`, i.e., it's preferable to write what you got and then what you expected.

e-dard commented:
@jonseymour Now I see. It wasn't clear to me that this issue relied on a successful call to `NewPointFromBytes` and then a call to `p.Fields`...

OK, just a small nit on the test case then squash all these down?

jonseymour commented:
Thanks for the review, @e-dard.

With this series, I've removed the test that used AddField and fixed the failure messages per the standard.

jonseymour commented:
Actually, I should check that NewPointFromBytes returns an error if there are no fields in the parsed point. I'll push a new series with those tests and changes.

jonseymour commented:
Ok, @e-dard  added two additional commits which require that points.NewFieldFromBytes() errors if a corrupt binary point does not yield at least one field.

It is quite late where I am (03:45), so I'll look for responses later in my morning.

jonseymour commented:
@e-dard, @jwilder - I have now extended this series of commits to ensure that the fields that result from a SELECT INTO query are unique and non-empty. The test case in 3518ae8 shows the behaviour.

Since this series now touches the query executor it would be worth getting @benbjohnson to have a look at it.

The RHS of the final merge could (in principle) be merged with the 0.10.0 maintenance branch (not suggesting that you do this, only that it is possible). The conflicts between the two branches were resolved in the merge commit.

jonseymour commented:
@jwilder I see that is an overlap with #5730. I am not sure if #5730 is guarantees uniqueness of column names. Does it need to? 

Do you want me to extend the changes to points.NewPoint()/points.NewPointFromBytes() to check that field names are unique?

jsternberg commented:
@jonseymour #5730 doesn't do anything with column uniqueness, it just gives a name to columns with a binary expression that is missing in the new query engine.

jonseymour commented:
Thanks for the info @jsternberg. 

I can see one argument for pushing up the column uniqueness logic from cluster/ to influxdb/ - it comes to the predictability of column names to be used by a SELECT INTO statement. As implemented above, the generated column names will be different to the names that appear in the influx UI, since for example, ```select max(power), max(temperature) from "device" group by time(1h)``` will currently get two max columns in the UI. With the change above, the columns will be max, and max_2, however without running the INTO query and inspecting the target, the user won't necessarily be able to predict that.

If the column uniqueness functionality is pushed up into influxql, then no downstream client of influxql (including the SELECT INTO logic) will have to worry about having to use their own algorithm to ensure unique column names which, I think, would be a very good thing. 

I'll await @ed-dard/@jwilder's comments about this before I make any further changes in this area.

jsternberg commented:
Whatever the naming scheme should be, the logic should probably go in `(*ast.Field).Name()` or `(*ast.SelectStatement).ColumnNames()`. That way a third party could use it to determine the name of the column directly from the select statement.

jsternberg commented:
@jwilder @e-dard looking at the code, it may be easier to just throw an error when a select statement is used that has conflicting output names and advise that the user uses `AS` to disambiguate rather than come up with some default collision detection that will work for everyone. We can also gradually improve the automated naming. A call to `max(value)` should probably be named `max_value` instead.

e-dard commented:
@jwilder should make the final call on this. 

Personally I would vote for generated column names that match the expression, e.g., `max(value)` would generate a column name `"max(value)"`, and for arbitrary math some placeholder, e.g., `col`, `col (1)`, `col (2)` or whatever.

jsternberg commented:
@e-dard I mentioned us changing how the column is generated, but we also need a way to resolve conflicts. When resolving conflicts we have three options.

* Automatically rename columns that are in conflicts
* Always name columns in a way that they can never possibly conflict
* Send an error back to the user and tell them to use an alias
* Do nothing (causes hidden errors)

The first one I think may be unintuitive, especially when it comes to `SELECT INTO` calls. Someone may do `SELECT value, value INTO mem FROM cpu` and be confused that suddenly it has implicitly renamed `value` to `value_0`. Maybe we can only rename later entries instead so the above call would be `value` and `value_1` and that would be more intuitive? That may be my favorite if we're going to do any implicit renaming.

Always naming columns different would cause confusion with a single value so I don't like it. It would cause `SELECT value FROM cpu` to output a column of `value_0` or `value_1` depending on the name.

The third is the safest, but causes existing queries to break and requires us to do a better job of choosing a better implicit name. It may not be feasible at the moment.

jonseymour commented:
@jsternberg I don't think this: 

> Someone may do SELECT value, value INTO mem FROM cpu and be confused that suddenly it has implicitly renamed value to value_0.

is necessarily a big problem if the UI produces the same result - at least the UI and SELECT INTO would produce the same result.

Personally my vote is with option 1 (deterministically rename), followed by option 3 (error out).

jon.

jsternberg commented:
@jonseymour I did a hangout with @jwilder and @e-dard to work through some of the use cases that have been discussed in this thread.

The implementation should probably be done in `ColumnNames()`. We didn't talk about that, but I think that would be a better place. It would allow someone to use the `influxql` package and call `ColumnNames()` to get the real column names without having to worry that they get implicitly renamed later. I don't yet know if this causes an issue for other places in the code though that may rely on `ColumnNames()` not doing that, so that's tentative assuming the implementation works.

We also discussed heavily how renames should be done and how aliases should be treated. We agreed on a two phase method of assigning names. First, aliases would get assigned to columns. Conflicts would not be implicitly renamed.

For other names, they would be resolved in the order they are present in the query with internal state being kept for which values had been encountered. For example, with `SELECT value, value_1, value + 2 FROM cpu`, you would get the columns `value, value_1, value_1_1`. This simple implementation would prevent names from conflicting with each other accidentally and also prevent us from having to do a very complicated set of recursions to determine how to resolve the names. Users could use aliases if they don't like the output.

I'm going to have a PR up later today with this implementation. Thank you for your contribution and for helping us to hash this out.

jonseymour commented:
Thanks @jsternberg that all makes sense. I've removed the uniqueness logic from this series.

@e-dard: I think I am about done this PR so I'll leave it with you for final review.

jsternberg commented on models/points.go:
```diff
@@ -25,6 +26,8 @@ var (
 		'=': []byte(`\=`),
 	}
+
+	ErrPointMustHaveAField = errors.New("Point without fields is unsupported")
```
I think we use lowercase text for our other errors.

jonseymour commented:
@e-dard I have squashed most of the commits in this series. There is one commit for the tests showing the problems that needed to be fixed and another for the fixes. The RHS of the merge commit merges cleanly with the 0.10.0 branch (which is in the form I need it for my production system). I have added a CHANGELOG update.

jonseymour commented:
I've removed the CHANGELOG.md update until I get a +1.

jwilder commented on models/points.go:
```diff
@@ -1079,6 +1081,9 @@ func NewPoint(name string, tags Tags, fields Fields, time time.Time) (Point, err
 			}
 		}
+		if len(key) == 0 {
+			return nil, fmt.Errorf("All fields must have non-empty names")
```
This error text should start be all lowercase.

jwilder commented:
One small thing, but :+1: otherwise.

jonseymour commented:
@jwilder - done. Thanks for the review!

e-dard commented:
LGTM. I'll merge this Monday morning @jonseymour

jonseymour commented:
Thank you, @e-dard
